src/app/content/highlights/reducer.ts

```diff
@@ -47,27 +61,104 @@ const reducer: Reducer<State, AnyAction> = (state = initialState, action) => {
         return state;
       }
 
-      return {
-        ...state,
-        highlights: state.highlights.map((highlight) =>
-          highlight.id === action.payload.id ? {
-            ...highlight,
-            ...action.payload.highlight,
-            color: action.payload.highlight.color as string as HighlightColorEnum,
-          } : highlight
-        ),
+      const { color, annotation } = action.payload.highlight;
+
+      const newState = {...state};
+      const dataToUpdate = {
+        ...action.payload.highlight,
+        color: action.payload.highlight.color as string as HighlightColorEnum,
       };
+      let newHighlight: Highlight | undefined;
+
+      newState.highlights = newState.highlights!.map((h) => {
+        if (h.id === action.payload.id) {
+          newHighlight = {
+            ...h,
+            ...dataToUpdate,
+          };
+          return newHighlight!;
+        }
+        return h;
+      });
+
+      const { chapterId, pageId } = action.meta;
+      const { filters: { chapters, colors }, highlights } = newState.summary;
+
+      // If highlight's chapter is not in summary filters stop here...
+      if (!chapters.includes(chapterId)) { return newState; }
+
+      // If only annotation was changed just update summary highlights...
+      if (!color && annotation) {
```

Comment:
all this branch and bound logic is nuts and should be extracted to a helper method if it can't be simplified

Reply:
Those are for cases which I can see when updating highlight which may be or may not be in summary highlights. I've moved logic outside reducer